## 10norootpassword.dpatch by Martin Pitt <martin.pitt@ubuntu.com>
##
## DP: When locking a root-owned screen, check that root has a
## DP: password set. If not, ask for an unlocking key.
## DP: See https://bugs.launchpad.net/bugs/6760

@DPATCH@
diff -Naur screen-4.0.3.orig/attacher.c screen-4.0.3/attacher.c
--- screen-4.0.3.orig/attacher.c	2008-06-01 23:39:48.000000000 +0200
+++ screen-4.0.3/attacher.c	2008-06-01 23:40:21.000000000 +0200
@@ -819,11 +819,15 @@
 #ifdef USE_PAM
   pam_handle_t *pamh = 0;
   int pam_error;
-#else
-  char *pass, mypass[16 + 1], salt[3];
 #endif
+  char *pass, mypass[16 + 1], salt[3];
+  int using_pam = 1;
 
-#ifndef USE_PAM
+#ifdef USE_PAM
+  if (!ppp->pw_uid)
+    {
+#endif
+  using_pam = 0;
   pass = ppp->pw_passwd;
   if (pass == 0 || *pass == 0)
     {
@@ -856,6 +860,8 @@
       pass = crypt(mypass, salt);
       pass = ppp->pw_passwd = SaveStr(pass);
     }
+#ifdef USE_PAM
+    }
 #endif
 
   debug("screen_builtin_lck looking in gcos field\n");
@@ -885,6 +891,8 @@
           AttacherFinit(SIGARG);
           /* NOTREACHED */
         }
+      if (using_pam)
+        {
 #ifdef USE_PAM
       PAM_conversation.appdata_ptr = cp1;
       pam_error = pam_start("screen", ppp->pw_name, &PAM_conversation, &pamh);
@@ -895,10 +903,13 @@
       PAM_conversation.appdata_ptr = 0;
       if (pam_error == PAM_SUCCESS)
 	break;
-#else
-      if (!strncmp(crypt(cp1, pass), pass, strlen(pass)))
-	break;
 #endif
+        }
+      else
+	{
+          if (!strncmp(crypt(cp1, pass), pass, strlen(pass)))
+	    break;
+	}
       debug("screen_builtin_lck: NO!!!!!\n");
       bzero(cp1, strlen(cp1));
     }
